src/ether_nethub.c

```diff
@@ -0,0 +1,599 @@
+
+#include <string.h>
+#include <unistd.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <sys/poll.h>
+#include <signal.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <netdb.h>
+#include <netinet/in.h>
+
+#include "commondefs.h"
+#include "lispemul.h"
+#include "lispmap.h"
+#include "emlglob.h"
+#include "lsptypes.h"
+#include "lspglob.h"
+#include "adr68k.h"
+#include "ether.h"
+#include "dbprint.h"
+#include "etherdefs.h"
+#include "ifpage.h"
+#include "iopage.h"
+#include "byteswapdefs.h"
+#include "dbprint.h"
+#include "timerdefs.h"
+
+/*
+**  --- ether implementation common data -------------------------------------------
+*/
+
+extern int      ether_fd;      /* file descriptor for ether socket */
+extern u_char   ether_host[6]; /* 48 bit address of this node */
+extern u_char   broadcast[6];
+extern int      ether_bsize;   /* if nonzero then a receive is pending */
+extern u_char  *ether_buf;     /* address of receive buffer */
+extern LispPTR *PENDINGINTERRUPT68k;
+extern fd_set   LispReadFds;
+
+extern int ETHEREventCount;
+
+/*
+**  --- nethub configuration data --------------------------------------------------
+*/
+
+static char* nethubHost = NULL;
+static int   nethubPort = 3333;
+static int   mac0 = 0xCA;
+static int   mac1 = 0xFF;
+static int   mac2 = 0xEE;
+static int   mac3 = 0x12;
+static int   mac4 = 0x34;
+static int   mac5 = 0x56;
+
+void setNethubHost(char* host) {
+  if (host && *host) {
+    nethubHost = host;
+    DBPRINT(("nh :: host now: '%s'\n", nethubHost));
+  }
+}
+
+void setNethubPort(int port) {
+  if (port > 0 && port <= 65535) {
+    nethubPort = port;
+    DBPRINT(("nh :: port now: %d\n", nethubPort));
+  }
+}
+
+void setNethubMac(int m0, int m1, int m2, int m3, int m4, int m5) {
+  if (   m0 >= 0 && m0 <= 255
+      && m1 >= 0 && m1 <= 255
+      && m2 >= 0 && m2 <= 255
+      && m3 >= 0 && m3 <= 255
+      && m4 >= 0 && m4 <= 255
+      && m5 >= 0 && m5 <= 255) {
+    mac0 = m0;
+    mac1 = m1;
+    mac2 = m2;
+    mac3 = m3;
+    mac4 = m4;
+    mac5 = m5;
+    DBPRINT(("nh :: mac now: %02X-%02X-%02X-%02X-%02X-%02X\n", mac0, mac1, mac2, mac3, mac4, mac5));
+  }
+}
+
+/*
+**  --- very simple logging --------------------------------------------------------
+*/
+
+static int loglevel = 0;
+static int logcnt = 0;
+
+void setNethubLogLevel(int ll) {
+  loglevel = ll;
+  DBPRINT(("nh :: loglevel now: %d\n", loglevel));
+}
+
+#define log_info(line) { if (loglevel == 1) \
+	{ printf("nh-info[%d]: ", logcnt++); printf line; }}
+#define log_debug(line) { if (loglevel >= 2) \
+	{ printf("nh-debug[%d]: ", logcnt++); printf line; }}
+#define log_all(line) { if (loglevel < 1) \
+	{ printf("nh[%d]: ", logcnt++); printf line; }}
+
+/*
+**  --- utilities ------------------------------------------------------------------
+*/
+
+#define LONG_PTR_MASK ((long)(~((long)3)))
+
+static void dblwordsSwap(u_char* basePtr, int forBytes) {
+  u_char* wordsPtr = (u_char*)((long)basePtr & LONG_PTR_MASK);
```

Comment:
See inc/lsptypes.h WORDPTR and the access macros defined for regular and BYTESWAP cases.

Reply:
no, `WORDPTR()` won't be helpful here:

- byte-swapping must start at a double-word boundary here and reverse the order of all 4 bytes in the double-words
- in the BYTESWAP-case `WORDPTR()` "simply" inverts the 2nd-to-last bit in the address, i.e. lets the address point to "the other" 16 bit word in a double-word
- this works only if the address points into the middle of a double-word: should the address point to the beginning of a double-word, all goes wrong
- i don't know if it is guaranteed that the packet buffer address will always start in the middle of a double-word
- and i still need the double-word address to compute the correct double-word count to byte-swap

So i left this procedure mostly unchanged and just renamed the variables to clarify that it's about double-words.

The only confusing thing left ist the name of the procedure `word_swap_page` which (despite its name) works on 32 bit chunks (reversing the bytes in double-words and not 16 bit words) and (worse) allows to start at the 2nd word (i.e. in the middle) of a double-word. (confusing in my opinion only, but tested with the mentioned byte-swapping outcomes)